src/main/kotlin/ch/srgssr/pillarbox/qos/event/SetupService.kt

```diff
@@ -0,0 +1,147 @@
+ * @property resourceLoader Loads resources such as JSON files for OpenSearch index templates and ISM policies.
+ * @property terminationService Responsible for terminating the application in case of failure.
+ * @property sseClient The client used to connect to the SSE endpoint once OpenSearch setup is complete.
+ */
+@Service
+class SetupService(
+    private val properties: OpenSearchConfigurationProperties,
+    private val resourceLoader: ResourceLoader,
+    private val terminationService: TerminationService,
+    private val sseClient: SseClient,
+) {
+    private val webClient: WebClient = WebClient.create(properties.uri.toString())
+
+    private companion object {
+        /**
+         * Logger instance for logging within this service.
+         */
+        private val logger = logger()
+
+        /**
+         * Path for creating the OpenSearch index.
+         */
+        private const val INDEX_CREATION_PATH = "/actions-000001"
+
+        /**
+         * Path for the Index State Management (ISM) policy in OpenSearch.
+         */
+        private const val ISM_POLICY_PATH = "/_plugins/_ism/policies/actions_policy"
+    }
+
+    /**
+     * Starts the OpenSearch setup process. This method checks the health of the OpenSearch cluster,
+     * applies the ISM policy, and creates the index if necessary. Once all setup tasks are complete,
+     * it starts the SSE client to begin receiving events.
+     */
+    fun start() {
+        checkOpenSearchHealth()
+            .retryWhen(
+                properties.retry.create().doBeforeRetry {
+                    logger.info("Retrying OpenSearch health check...")
+                },
+            ).doOnSuccess { logger.info("OpenSearch is healthy, proceeding with setup...") }
+            .then(checkAndApplyISMPolicy())
+            .then(checkAndCreateIndex())
+            .doOnSuccess { logger.info("All setup tasks are completed, starting SSE client...") }
+            .subscribe(
+                { sseClient.start() },
+                { CoroutineScope(Dispatchers.IO).launch { terminateApplication(it) } },
+            )
+    }
+
+    private fun terminateApplication(error: Throwable) {
+        logger.error("Failed to connect to OpenSearch:", error)
+        terminationService.terminateApplication()
+    }
+
+    // Check OpenSearch health using WebClient
+    private fun checkOpenSearchHealth(): Mono<ResponseEntity<Void>> =
+        webClient
+            .get()
+            .uri("/")
+            .retrieve()
+            .toBodilessEntity()
+
+    private fun checkAndCreateIndex(): Mono<ResponseEntity<Void>> =
+        webClient
+            .head()
+            .uri(INDEX_CREATION_PATH)
+            .retrieve()
+            .onStatus({ status -> status.is4xxClientError }) {
+                logger.info("Index does not exist, creating index...")
+                createIndex().then(Mono.empty())
+            }.onStatus({ status -> status.is2xxSuccessful }) {
```

Comment:
These two statuses could also use method reference:
```suggestion
            .onStatus(HttpStatusCode::is4xxClientError) {
                logger.info("Index does not exist, creating index...")
                createIndex().then(Mono.empty())
            }.onStatus(HttpStatusCode::is2xxSuccessful) {
```